src/quick-lint-js/parse.h

```diff
@@ -2515,8 +2580,42 @@ class parser {
     }
 
     if (this->peek().type == token_type::kw_else) {
+      source_code_span else_span = this->peek().span();
       this->skip();
-      parse_and_visit_body();
+
+      switch (this->peek().type) {
+      default:
+        parse_and_visit_body();
+        break;
+
+      case token_type::left_paren:
+        expression *ast = this->parse_expression(precedence{});
+        this->visit_expression(ast, v, variable_context::rhs);
+
+        bool is_invalidating_if = false;
+
+        switch (ast->kind()) {
+        default:
+          break;
+
+        // Any other kind?
+        case expression_kind::arrow_function_with_expression:
+        case expression_kind::arrow_function_with_statements:
+          is_invalidating_if = true;
+          break;
+        }
+
+        if (this->peek().type == token_type::left_curly) {
+          if (!is_invalidating_if && this->is_side_effect_free(ast)) {
+            this->error_reporter_->report(
+                error_else_with_conditional_missing_if{
+                    .else_token = else_span,
+                });
+          }
+          parse_and_visit_body();
+        }
+        break;
```

Comment:
Blocking: After parsing the expression statement, the language says we need a semicolon. (A semicolon might be inserted by ASI if a newline or `}` is found instead of a semicolon.)

In other words, the following code should produce a syntax error (and quick-lint-js does report a syntax error before your commit):
```js
if (a) { b; } else (() => console.log(c))() { d; }
```